### PR TITLE
12 Ratings Widget

### PR DESCRIPTION
**Pull request by DurtDiv3r**
**Work Done:**
Add colours for Rating Widget stars.
Rename composable to RatingBar.
Refactor calculateRating.
Update Constants BASE_URL and add path for star.
Implement custom Rating Widget as no implementation for Jetpack Compose yet.

**Reference:**
Rating example:
https://www.jetpackcompose.app/snippets/RatingBar

Custom colors:
https://developer.android.com/jetpack/compose/designsystems/custom

**Screenshot:**
RatingWidget

### CODE REVIEW

File: app/src/main/java/com/islaharper/dawnofandroid/presentation/components/RatingWidget.kt [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+import com.islaharper.dawnofandroid.ui.theme.DarkStarBackgroundColour
+import com.islaharper.dawnofandroid.ui.theme.DarkStarColour
+import com.islaharper.dawnofandroid.ui.theme.DawnOfAndroidTheme
+import com.islaharper.dawnofandroid.ui.theme.LightStarBackgroundColour
+import com.islaharper.dawnofandroid.ui.theme.LightStarColour
+import com.islaharper.dawnofandroid.ui.theme.PADDING_SMALL
+import com.islaharper.dawnofandroid.util.Constants.STAR_PATH
+
+@Composable
+fun RatingWidget(
```
Comment (drawers) [inline app/src/main/java/com/islaharper/dawnofandroid/presentation/components/RatingWidget.kt]: Thanks for this. Is this based on a sample somewhere e.g., in a course or in an article or Stack Overflow? If it is, please add a link or PDF if paywalled. This is really important to help reviewers understand the full context.

Comment (DurtDiv3r) [reply app/src/main/java/com/islaharper/dawnofandroid/presentation/components/RatingWidget.kt]: Hi, I looked at https://www.jetpackcompose.app/snippets/RatingBar as a starting point and implemented something I was happy with.  I implemented a similar solution in the FNB App for a rating widget using the path from an SVG a while back. Don't have an example of previous java implementation as it's in the FNB codebase but I've attached a screenshot from it's current use in the FNB app.

rating

File: app/src/main/java/com/islaharper/dawnofandroid/domain/model/CustomColourScheme.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import com.islaharper.dawnofandroid.ui.theme.LightStarBackgroundColour
+import com.islaharper.dawnofandroid.ui.theme.LightStarColour
+
+/* https://developer.android.com/jetpack/compose/designsystems/custom
+    Add additional custom colours as needed
+    Usage: DawnOfAndroidTheme.colourScheme.myColour
+*/
+
+@Immutable
+data class CustomColourScheme(
```
Comment (drawers) [inline app/src/main/java/com/islaharper/dawnofandroid/domain/model/CustomColourScheme.kt]: Maybe I'm wrong here, but I was thinking more like a solution like the following:

```kotlin
object RatingBarDefaults {

    @Composable
    fun colors(): RatingBarColors {
        return if (isSystemInDarkTheme()) {
            RatingBarColors(
                starColor = Color(0x),
                starBackgroundColor = Color(0x)
            )
        } else {
            RatingBarColors(
                starColor = Color(0x),
                starBackgroundColor = Color(0x)
            )
        }
    }
}

data class RatingBarColors(
    val starColor: Color,
    val starBackgroundColor: Color
)

@Composable
fun RatingBar(
    rating: Double,
    modifier: Modifier = Modifier,
    spaceBetween: Dp = PADDING_SMALL,
    colors: RatingBarColors = RatingBarDefaults.colors()
) {
```

File: app/src/main/java/com/islaharper/dawnofandroid/presentation/components/RatingBar.kt [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+import com.islaharper.dawnofandroid.ui.theme.PADDING_SMALL
+import com.islaharper.dawnofandroid.util.Constants.STAR_PATH
+
+@Composable
+fun RatingBar(
+    rating: Double,
+    modifier: Modifier = Modifier,
+    spaceBetween: Dp = PADDING_SMALL
+) {
+    val result = calculateRating(rating = rating)
```
Comment (drawers) [inline app/src/main/java/com/islaharper/dawnofandroid/presentation/components/RatingBar.kt]: We don't want `calculateRating(rating = rating)` to be executed every time. I think it should be something like:

```kotlin
    val result by remember(rating) { mutableStateOf(calculateRating(rating = rating)) }
```

File: app/src/main/java/com/islaharper/dawnofandroid/presentation/components/RatingBar.kt [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+                starBackgroundColour = DawnOfAndroidTheme.colourScheme.starBackgroundColour
+            )
+        }
+    }
+}
+
+@Preview(name = "Light", showBackground = true)
+@Preview(name = "Dark", uiMode = UI_MODE_NIGHT_YES, showBackground = true)
+@Composable
+fun RatingBarPreview() {
```
Comment (drawers) [inline app/src/main/java/com/islaharper/dawnofandroid/presentation/components/RatingBar.kt:69]: Thanks, these previews are better now

**Review by drawers — APPROVED** [review]

File: app/src/main/java/com/islaharper/dawnofandroid/domain/model/CustomColourScheme.kt (hunk 2 above)
Comment (DurtDiv3r) [reply app/src/main/java/com/islaharper/dawnofandroid/domain/model/CustomColourScheme.kt]: My initial thinking was that additional colours could be added if needed and accessed in a similar way to MaterialTheme.colorScheme, but I do like that your solution provides better context in the composable for the colours used.

**Pull request merged**